x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java

```diff
@@ -83,25 +86,24 @@ public SecurityActionFilter(AuthenticationService authcService, AuthorizationSer
         if (licenseState.isSecurityEnabled()) {
             final ActionListener<Response> contextPreservingListener =
                     ContextPreservingActionListener.wrapPreservingContext(listener, threadContext);
+            final ActionListener<Response> postActionExecutionListener = ActionListener.delegateFailure(contextPreservingListener,
+                    (ignore, response) -> {
+                        String requestId = AuditUtil.extractRequestId(threadContext);
+                        Authentication authentication = securityContext.getAuthentication();
```

Comment:
I haven't mentioned this as a drawback when we discussed it. I think I wasn't fully confident in my ability to clearly explain it. But this is part of the complexity and maintenance burden that I've mentioned, so I'll try to explain it here:

In order to pass the `requestId` and `authentication` as arguments to the new audit method, they need to be pulled from the thread context. Currently the `requestId` and `authentication` ARE present in the thread context when the listener is called with the response. But to ensure this forward we'll have to introduce an assertion (probably at this very place in the code). Even then, we cannot guarantee it because tests don't cover all the paths in an action handler: it could be that some handler stashes the thread context and calls the listener with a response (without calling another transport action, which is the usual course of action when stashing the context).

Alternatively, we could enforce it more stringently. Instead of an assert we could hard fail (reroute the call to `listener.onResponse` to `listener.onFailure`) if we detect that the thread context has been altered.

I favour the latter, but maybe this is overdoing it since the "danger" is inaccurate (missing `requestId` and `authentication`) audit logging for action responses (for which there is not concrete use case atm, but think of auditing creating API keys as an example).

We could also opt for assets in the minor and hard fail in the major release, similar to what we've discussed for indices privileges.

Reply:
> Alternatively, we could enforce it more stringently. Instead of an assert we could hard fail (reroute the call to listener.onResponse to listener.onFailure) if we detect that the thread context has been altered.

I've pushed https://github.com/elastic/elasticsearch/pull/63708/commits/2f5e54682ab7abfb8f8e4c47d9b1dd5f70ab8503
 to better illustrate what I mean.

Reply:
I've iterated on the issue, this is no longer a problem.